### DISCUSSION TITLE
How to interpret error objects?

### DISCUSSION THREAD

**Discussion by adam-rgare** [sql-formatter-org/sql-formatter#595] [Q&A] [answered]
This error was thrown when attempting to format an improper query:

```
{
    "offset": 28,
    "token": {
        "type": "RESERVED_CLAUSE",
        "raw": "FROM",
        "text": "FROM",
        "start": 155,
        "precedingWhitespace": "\n "
    }
}

```

I am highlighting the text so the user who wrote the query can see where in the SQL their syntax error occurred.

I originally assumed that the error would be in `exampleSQL.slice(token.start, token.start + offset)`, but it actually appears in `exampleSQL.slice(token.start - offset, token.start)`.

Is this expected behavior? Asking so I don't end up introducing new bugs to our application with some out of range errors in the future.

**Comment by nene** [comment, marked as answer]
Hard to reproduce without the exact code that triggers this error, but I suspect this object comes from the parser which in turn refers to the tokens created by the Tokenizer. I personally haven't seen this kind of object being thrown. Normally I would expect some instance of `Error` class to be thrown, not a plain object.

The `token.start` property refers to the position in source text where this token starts, in this case the start of "FROM". Where this `offset` comes from... I have no idea. Perhaps it's the offset of the current token in the array of tokens.

Anyway, to my knowledge there's no clear start/end info inside tokens. So the best you can do is something like:

```
start = token.start
end = token.start + token.raw.length
```

Note. It might seem to you that the actual syntax error happened earlier. But from the parser perspective the location might be different. Here I think the problem was that the parser encountered FROM token in a place where it didn't expect it.

**Reply by adam-rgare** [reply]
> Here I think the problem was that the parser encountered FROM token in a place where it didn't expect it.

Seems like common sense in retrospect! Thanks for your help, @nene.

Only thing is, the token start was not at the 'F' in FROM, but a few arbitrary letters before that.

**Reply by nene** [reply]
Perhaps a difference between counting bytes v/s characters.
